libs/utils: add frequency plotting for peripheral clocks

kdub opened the pull request:
Allows the plotting of clocks using the linux common clock tracing
infrastructure tracing. Chart shows the periods the clock is on or
off, as well as frequency information.

Test: run examples/android/workloads/Android_YouTube.ipynb, which
Test: now has a plotPeripheralFrequencies() example

kdub commented:
new notebook: (not viewed well in diff)
https://gist.github.com/kdub/6e8f0e064d37ca24c493edc649df5dc8

mdigiorgio commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -115,6 +115,127 @@ def _dfg_cluster_frequency_residency(self, cluster, total=True):
+          enable_df['clock_setting'] = 1;
+        if not disable_df.empty:
+          disable_df = disable_df[disable_df.clk_name == clk]
+          disable_df['clock_setting'] = 0;
```
Different indentation

derkling commented:
@kdub, any chance you can update this series?

kdub commented:
Sure, will work on updating now, sorry for delay.

kdub commented:
Adopted changes, thanks for the suggestion!

kdub commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -115,6 +115,127 @@ def _dfg_cluster_frequency_residency(self, cluster, total=True):
+        disable_df = self._dfg_trace_event('clock_disable')
+        pd.options.mode.chained_assignment = None
+
+        rate_df = rate_df[rate_df.clk_name == clk]
```
done.

valschneider commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +163,127 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+        rate_df = self._dfg_trace_event('clock_set_rate')
+        enable_df = self._dfg_trace_event('clock_enable')
+        disable_df = self._dfg_trace_event('clock_disable')
+        pd.options.mode.chained_assignment = None
```
Usually when we face SettingWithCopy warnings we try to quiet them down by using `df.(i)loc ...`, even if the df is being modified properly. 

I see that this is isn't followed in plotClusterFrequencies(), but the warning is eventually restored with `pd.options.mode.chained_assignment = 'warn'`. If we really can't change the code to not have the warning I'd be fine with temporarily disabling it, but then it should be restored eventually.

derkling replied:
Likelt plotClusterFrequencies is warning because the pandas library added that check after that function as been added... but, going forward, I would say we should try to avoid warnings instead of muting them.

kdub replied:
Changed to not suppress warning, and corrected offending line to not generate warning.

valschneider commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +163,127 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+            self._log.warning('No events for clock ' + clk + ' found in trace')
+            return
+
+        #this will extend the last frequency value to the end of the timerange.
```
This is top-level nitpicking, but... Could you please insert a space between `#` and your comments ? Bonus points if they start with a capital letter too

kdub replied:
done

valschneider commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +163,127 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+        for index, row in freq.iterrows():
+            if row.state == -1 :
+                last_freq = row.clock_setting
+                if max_freq < last_freq: max_freq = last_freq
```
Can't we do `max_freq = rate_df.clock_setting.max()` outside of the loop ?
I wonder if this whole loop can't be done with some of pandas' clever functions like `applymap()` to set the frequency to 0 when `clock_setting == 0`, else keep it. The last row logic could then be handled after that.

derkling replied:
:+1: for the apply approach

kdub replied:
I changed it to use max(), but I didn't find it easy to use the apply() or applymap() approach. I always had to put in more state (namely the frequency value of the  last row) than I could easily capture in the lambda.

valschneider commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +163,127 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+        state_axis.set_xlim(self._trace.x_min, self._trace.x_max)
+        state_axis.legend(bbox_to_anchor=(0., 1.02, 1., 0.102), loc=3, ncol=3, mode='expand')
+
+        figname = '{}/{}{}.png'\
```
`os.path.join()` would be better.

kdub replied:
done, the other functions could use an update too as a followup.

joelagnel commented:
@kdub any updates on this PR?

kdub commented:
Thanks for taking another look, comments should be fixed.

valschneider commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +163,124 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+        disable_df = self._dfg_trace_event('clock_disable')
+
+        rate_df = rate_df[rate_df.clk_name == clk]
+        rate_df['clock_setting'] = rate_df['rate']/1e6
```
I get setting with copy warnings here. If this is needed for all clock-related analysis, this could be added to trace sanitization (in **libs/utils/trace.py**) and applied to all the traces (no filtering on clk).

Otherwise, if this is just needed for this analysis, you could do this:
```
rate_df = rate_df[rate_df.clk_name == clk].copy()
```

kdub replied:
Adopted changes, although I couldn't reproduce the warning here with my setup (pandas 0.22)

valschneider commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +163,124 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+
+        if not enable_df.empty:
+            enable_df = enable_df[enable_df.clk_name == clk]
+            enable_df['clock_setting'] = 1;
```
Ditto for the warning. See above comment or use:
```
enable_df.loc[:, 'clock_setting'] = 1;
```

kdub replied:
Adopted changes, although I couldn't reproduce the warning here with my setup (pandas 0.22)

valschneider commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +163,124 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+        max_freq = rate_df['clock_setting'].max()
+        last_state = 1
+        last_freq = rate_df.iloc[0]['clock_setting']
+        for index, row in freq.iterrows():
```
Had a look at the trace example in trappy, makes more sense now (I didn't know how the dataframes would be built).

It is indeed a tricky one. I couldn't think of a decent `apply()` use, but I think we can make it shorter and easier to read:

```
max_freq = rate_df['clock_setting'].max()
for index, row in freq.iterrows():
    # Get latest frequency update
    curr_freq = rate_df[:index]['clock_setting'].values[-1]
    freq.loc[index, 'frequency'] = 0 if not row.state else curr_freq
```

What do you think ?

kdub replied:
sure, adopted. Thanks!

valschneider commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +163,124 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+
+        # This will extend the last frequency value to the end of the timerange.
+        last = freq.tail(n=1)
+        last.index = { self._trace.x_max }
```
What about directly computing the delta of each event, and "fixing" the last one to extend it to the end of the trace ? This is what we do for overutilized:
```
df = self._dfg_trace_event('sched_overutilized')
df['start'] = df.index
df['len'] = (df.start - df.start.shift()).fillna(0).shift(-1)
df.drop('start', axis=1, inplace=True)

# Fix the last event, which will have a NaN duration
# Set duration to trace_end - last_event
df.loc[df.index[-1], 'len'] = self.start_time + self.time_range - df.index[-1]
```

kdub replied:
I don't mind adopting this if you insist, but left as-is, is computing the the delta as fast as just duplicating the last row at a different timestamp to fix up the chart?

derkling replied:
I like better to have code consistency, just because we can easy recognise similar code patterns and likely, maybe in the future, factorise them into a new common API.

valschneider commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +163,124 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+            enable_df['clock_setting'] = 1;
+        if not disable_df.empty:
+            disable_df = disable_df[disable_df.clk_name == clk]
+            disable_df['clock_setting'] = 0;
```
Ditto for the warning

kdub replied:
Adopted changes, although I couldn't reproduce the warning here with my setup (pandas 0.22).

kdub commented:
@valschneider 
* I refactored the code so the dataframe should be reusable (in the same manner as getCPUActiveSignal). 
* I also reworked the delta computing code to be more like getCPUActiveSignal.
* I added a unit test as well for trace parsing.
* I also regenerated the notebook: https://gist.github.com/kdub/6e8f0e064d37ca24c493edc649df5dc8

kdub commented:
Anyone have time to review again? Thanks!

joelagnel commented:
It sad github shows "Changes requested" when changes were already made. Why does github not get it? Atleast they should remove the red from Changes requested since it completely misleads. :-(

Kevin I'll look at this shortly! Thanks.

joelagnel commented:
LGTM. @derkling  ?

valschneider commented on tests/lisa/test_trace.py:
```diff
@@ -245,6 +246,37 @@ def test_sched_load_avg_signals(self):
+          <idle>-0 [004] 380339000000: cpu_idle:             state=1 cpu_id=4
+        """)
+        df = trace.getPeripheralClockEffectiveRate(clk_name='bus_clk')
+        print(df)
```
Eh, could have used that trace to test my previous comments ;_;
Anyway, tests are cool, so thanks for adding one. I'd just like the print to go please.

kdub commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +163,86 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+        plt.suptitle(title, y=.97, fontsize=16, horizontalalignment='center')
+
+
+        #plot frequency information (set rate)
```
fixed

kdub commented:
@valschneider Fixed issues found, using ffill was a much simpler solution, thanks!

valschneider commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +164,85 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+        plt.suptitle(title, y=.97, fontsize=16, horizontalalignment='center')
+
+
+        # plot frequency information (set rate)
```
Capitalisation please

valschneider commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +164,85 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+        else:
+            self._log.warning('No clock_set_rate events to plot')
+
+        # plot frequency information (effective rate)
```
Ditto capitalisation

valschneider commented on libs/utils/analysis/frequency_analysis.py:
```diff
@@ -163,6 +164,85 @@ def _dfg_cpu_frequency_transition_rate(self, cpu):
+                          linewidth = 10.0, label='clock off', color='red')
+
+
+        #plot time period that the clock state was unknown from the trace
```
Space + capitalisation

valschneider commented on libs/utils/trace.py:
```diff
@@ -874,6 +874,39 @@ def getClusterActiveSignal(self, cluster):
+
+        freq['start'] = freq.index
+        freq['len'] = (freq.start - freq.start.shift()).fillna(0).shift(-1)
+        #the last value will be NaN, fix to be appropriate length
```
Space + capitalisation

kdub commented:
@valschneider comments and commit ordering fixes done.
